iwyu.cc

```diff
@@ -1570,13 +1572,14 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
 
   // Called when the given type is fully used at used_loc, regardless
   // of the type being explicitly written in the source code or not.
-  virtual void ReportTypeUse(SourceLocation used_loc, const Type* type) {
-    ReportTypeUseInternal(used_loc, type, blocked_types_);
+  virtual void ReportTypeUse(SourceLocation used_loc, const Type* type,
+                             DerefKind deref_kind) {
+    ReportTypeUseInternal(used_loc, type, blocked_types_, deref_kind);
   }
 
   void ReportTypesUse(SourceLocation used_loc, const set<const Type*>& types) {
     for (const Type* type : types)
-      ReportTypeUse(used_loc, type);
+      ReportTypeUse(used_loc, type, DerefKind::None);
```

Comment:
Just out of curiosity, could this ever need per-type deref kind? Or are batch-reported types always non-ptr/ref by some implicit rule?

Reply:
(definitely not saying we should do anything about it now, just curious)

Reply:
Yes, an implicit rule that `ReportTypesUse` is called for reporting instantiated template cached types which are stored already dereferenced according to original `DerefKind`s.

But it seems for me quite natural that `DerefKind` is `None` when is not explicitly specified. If ever needed, `ReportTypesUse` interface should be changed to accept `DerefKind` for all the types or for every type.

Reply:
Makes sense, thanks!